compiler-core/src/pretty.rs

```diff
@@ -218,6 +218,22 @@ pub enum NestCondition {
     IfBroken,
 }
 
+/// Used to change the way nesting of documents work.
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub enum NestMode {
```

Comment:
Why is this needed? I couldn't figure out why one couldn't set it back to zero but have the normal indent rules still apply.

Reply:
I'm not sure I get the question right, but this change is needed because before adding this enum everything implicitly worked in the `Increase` mode. Indentation could only be increased, with `nest(doc, 2)` we're saying increase the indentation level by 2 but we have no way to know how much the document is actually going to be indented because it might depend _where_ the document is used.

```rust
// This is pseudo syntax
let doc = group(["foo", break(" ", ""), "bar"])

doc.nest(2).to_string(1)
// foo
//   bar
// ^^---- `.nest(2)` increased the nesting by 2

doc.nest(2).nest(0).to_string(1)
// foo
//   bar
// ^^---- `.nest(0)` increases the nesting by 0
//        so it is a no-op, it doesn't set the
//        indentation back to zero

doc.nest(2).nest(-2).to_string(1)
// foo
// bar
// ^---- this is a hack one can use to reduce nesting
//       but it is unreliable as one would have to know
//       beforehand how much a document is going to be
//       nested
```

Multiline strings require that the indentation of each of their lines is always set to 0 because adding indentation would change the content of the multiline string. The set mode is a way to reliably set the indentation of a piece of document to an exact value:

```rust
doc.set_nesting(0).nest(2).to_string(1)
// foo
// bar
// ^---- no matter how many following nests, the
//       indentation of `doc` will always be 0
//       this is what we need for multiline strings:
//       an outer nest must not change the string's
//       content
```

Reply:
Sorry, I'm still finding this hard to understand.

Why is it that Elixir doesn't need this algebra extension to do this?

Reply:
Yes it needs it ([here](https://github.com/elixir-lang/elixir/blob/74bfab8ee271e53d24cb0012b5db1e2a931e0470/lib/elixir/lib/inspect/algebra.ex#L220)), it uses the atom `:reset` passed to the nest function and [it always puts nesting at 0](https://github.com/elixir-lang/elixir/blob/74bfab8ee271e53d24cb0012b5db1e2a931e0470/lib/elixir/lib/inspect/algebra.ex#L1044C12-L1044C12), it would be the same as doing `doc.set_nesting(0)` here.